db/migrate/20190906165501_remove_broken_variants_from_carts.rb

```diff
@@ -0,0 +1,15 @@
+class RemoveBrokenVariantsFromCarts < ActiveRecord::Migration
+  def up
+    # Removes line_items from open carts where the variant has a hard-deleted price
+
+    variants_without_prices = 'SELECT spree_variants.id FROM spree_variants
+      LEFT OUTER JOIN spree_prices ON (spree_variants.id = spree_prices.variant_id)
+      WHERE NOT EXISTS (SELECT 1 FROM spree_prices WHERE spree_prices.variant_id = spree_variants.id)'
```

Comment:
Wow, I've never seen this clause. I would have written this:
```sql
WHERE spree_prices.id IS NULL
```

In the joined table, all variants that don't have a price will have NULL in all spree_prices columns. @sauloperez What's your opinion on this?

Another way would be to use a subquery:
```sql
SELECT spree_variants.id FROM spree_variants
WHERE spree_variants.id NOT IN (SELECT spree_prices.id FROM spree_prices)
```
But I'm not sure what's more efficient. I believe that a subquery would be really slow in Mysql because it doesn't index that sub-query result. Not sure about Postgres.

Reply:
Yep, good catch @mkllnk . We need to filter out variants that have price with that `WHERE spree_prices.id IS NULL`. @Matt-Yorkley https://blog.codinghorror.com/a-visual-explanation-of-sql-joins/ is your friend.

Reply:
I think this is the better SQL:
```
variants_without_prices = 'SELECT spree_variants.id FROM spree_variants
      LEFT OUTER JOIN spree_prices ON (spree_variants.id = spree_prices.variant_id)
      WHERE spree_prices.id IS NULL
```

Reply:
I like the suggestion of @luisramos0. :+1:

Reply:
credits to you @mkllnk :trollface:

Reply:
These queries seem to give exactly the same results:
```ruby
SELECT spree_variants.id FROM spree_variants LEFT OUTER JOIN spree_prices ON (spree_variants.id = spree_prices.variant_id) 
WHERE NOT EXISTS (SELECT 1 FROM spree_prices WHERE spree_prices.variant_id = spree_variants.id)
```

```ruby
SELECT spree_variants.id FROM spree_variants LEFT OUTER JOIN spree_prices ON (spree_variants.id = spree_prices.variant_id) 
WHERE spree_prices.id IS NULL
```